sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,661 @@
+        )
+        .expect("Adding container name to key resulted in invalid format");
+        let value = serialize_scalar(&"off".to_string()).context(error::Serialize)?;
+
+        let mut map = HashMap::new();
+        map.insert(key, value);
+        let settings: model::Settings = datastore::deserialization::from_map(&map)
+            .context(error::SettingsDeserialize { settings: map })?;
+
+        info!("Turning off container '{}'", container_id);
+        apiclient::set::set(socket_path, &settings)
+            .await
+            .context(error::Set)?;
+    }
+
+    Ok(())
+}
+
+async fn run() -> Result<()> {
+    let (args, subcommand) = parse_args(env::args())?;
+
+    // SimpleLogger will send errors to stderr and anything less to stdout.
+    SimpleLogger::init(args.log_level, LogConfig::default()).context(error::Logger)?;
+
+    info!("bootstrap-containers started");
+
+    match subcommand {
+        Subcommand::CreateContainers => create_containers(args.socket_path).await,
+        Subcommand::MarkBootstrap(mark_bootstrap_args) => {
+            mark_bootstrap(mark_bootstrap_args, args.socket_path).await
+        }
+    }
+}
+
+// Returning a Result from main makes it print a Debug representation of the error, but with Snafu
+// we have nice Display representations of the error, so we wrap "main" (run) and print any error.
+// https://github.com/shepmaster/snafu/issues/110
+#[tokio::main]
+async fn main() {
+    if let Err(e) = run().await {
+        match e {
+            error::Error::Usage { .. } => {
+                eprintln!("{}", e);
+                usage();
+                process::exit(1);
+            }
+            _ => {
+                eprintln!("{}", e);
+                process::exit(1);
+            }
+        }
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+mod error {
+    use datastore::Key;
+    use snafu::Snafu;
+    use std::collections::HashMap;
+    use std::fmt;
+    use std::io;
+    use std::path::PathBuf;
+    use std::process::{Command, Output};
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility = "pub(super)")]
+    pub(super) enum Error {
+        #[snafu(display("Error sending {} to {}: {}", method, uri, source))]
+        APIRequest {
+            method: String,
+            uri: String,
+            source: apiclient::Error,
+        },
+
+        #[snafu(display("Unable to base64 decode user-data '{}': '{}'", base64_string, source))]
+        Base64Decode {
+            base64_string: String,
+            source: base64::DecodeError,
+        },
+
+        #[snafu(display("Invalid bootstrap container mode '{}'", mode))]
+        BootstrapContainerMode {
+            mode: String,
+            source: model::modeled_types::error::Error,
+        },
+
+        #[snafu(display("'{}' failed - stderr: {}",
+                        bin_path, std::str::from_utf8(&output.stderr).unwrap_or_else(|_| "<invalid UTF-8>")))]
```

Comment:
Probably better to display as best as we can rather than hide all of it.

```suggestion
        #[snafu(display("'{}' failed - stderr: {}", bin_path, output.stderr.to_string_lossy())]
```

Reply:
Updated the code with `String::from_utf8_lossy`